pkg/ingress/model_build_load_balancer.go

```diff
@@ -269,78 +269,120 @@ func (t *defaultModelBuildTask) buildLoadBalancerSubnetMappings(ctx context.Cont
+			}
 		} else {
-			backendSGID, err := t.backendSGProvider.Get(ctx, networking.ResourceTypeIngress, k8s.ToSliceOfNamespacedNames(t.ingGroup.Members))
+			frontendSGIDs, err := t.sgResolver.ResolveViaSelector(ctx, chosenSGSelector)
 			if err != nil {
 				return nil, err
 			}
-			t.backendSGIDToken = core.LiteralStringToken((backendSGID))
-			t.backendSGAllocated = true
-			lbSGTokens = append(lbSGTokens, t.backendSGIDToken)
-		}
-		t.logger.Info("Auto Create SG", "LB SGs", lbSGTokens, "backend SG", t.backendSGIDToken)
-	} else {
-		manageBackendSGRules, err := t.buildManageSecurityGroupRulesFlag(ctx)
-		if err != nil {
-			return nil, err
-		}
-		frontendSGIDs, err := t.sgResolver.ResolveViaNameOrID(ctx, sgNameOrIDsViaAnnotation)
-		if err != nil {
-			return nil, err
+			for _, sgID := range frontendSGIDs {
+				lbSGTokens = append(lbSGTokens, core.LiteralStringToken(sgID))
+			}
+			if chosenSGSelector.ManagedBackend != nil && *chosenSGSelector.ManagedBackend {
+				backendSGID, err := t.backendSGProvider.Get(ctx, networking.ResourceTypeIngress, k8s.ToSliceOfNamespacedNames(t.ingGroup.Members))
+				if err != nil {
+					return nil, err
+				}
+				t.backendSGIDToken = core.LiteralStringToken(backendSGID)
+				t.backendSGAllocated = true
+				lbSGTokens = append(lbSGTokens, t.backendSGIDToken)
+			}
+			return lbSGTokens, nil
 		}
-		for _, sgID := range frontendSGIDs {
-			lbSGTokens = append(lbSGTokens, core.LiteralStringToken(sgID))
+	}
+
+	if len(explicitSGNameOrIDsList) > 0 {
+		sgNameOrIDsViaAnnotation := explicitSGNameOrIDsList[0]
+		for _, sgNameOrIDs := range explicitSGNameOrIDsList[1:] {
+			if !cmp.Equal(sgNameOrIDsViaAnnotation, sgNameOrIDs) {
+				return nil, fmt.Errorf("conflicting securityGroups: %v | %v", sgNameOrIDsViaAnnotation, sgNameOrIDs)
+			}
 		}
 
-		if manageBackendSGRules {
-			if !t.enableBackendSG {
-				return nil, errors.New("backendSG feature is required to manage worker node SG rules when frontendSG manually specified")
+		if len(sgNameOrIDsViaAnnotation) > 0 {
+			manageBackendSGRules, err := t.buildManageSecurityGroupRulesFlag(ctx)
+			if err != nil {
+				return nil, err
 			}
-			backendSGID, err := t.backendSGProvider.Get(ctx, networking.ResourceTypeIngress, k8s.ToSliceOfNamespacedNames(t.ingGroup.Members))
+			frontendSGIDs, err := t.sgResolver.ResolveViaNameOrID(ctx, sgNameOrIDsViaAnnotation)
 			if err != nil {
 				return nil, err
 			}
-			t.backendSGIDToken = core.LiteralStringToken(backendSGID)
-			t.backendSGAllocated = true
-			lbSGTokens = append(lbSGTokens, t.backendSGIDToken)
-		}
-		t.logger.Info("SG configured via annotation", "LB SGs", lbSGTokens, "backend SG", t.backendSGIDToken)
-	}
-	return lbSGTokens, nil
-}
+			for _, sgID := range frontendSGIDs {
+				lbSGTokens = append(lbSGTokens, core.LiteralStringToken(sgID))
+			}
 
-func (t *defaultModelBuildTask) buildFrontendSGNameOrIDsFromAnnotation(ctx context.Context) ([]string, error) {
```

Comment:
I'm just thinking would it be more maintainable and readable if we have separate functions like `buildFrontendSGNameOrIDsFromAnnotation()` and `buildFrontendSGNameOrIDsFromSelector()` and call them in the `buildLoadBalancerSecurityGroups()`?

Reply:
Extracted two receivers.